### PR TITLE
feat:[CDS-50810]: implement terraform cloud connector

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Implementing a new connector for Terraform Cloud. Connection should be established with an API token.  
Connection validation is hard coded for now to always returns a successful connection until the API client is implemented
A New Delegate task type is added TERRAFORM_CLOUD_TASK_NG . Currently flagged as unsupported, until the rest of the feature is implemented and tested
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)

### CODE REVIEW

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed 43 of 47 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 47 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`440-connector-nextgen/src/main/java/io/harness/connector/mappers/terraformcloudmapper/TerraformCloudEntityToDTO.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDFksoBMpjdNpt6uEM:-NMDFksoBMpjdNpt6uEN:b-quszmh) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/440-connector-nextgen/src/main/java/io/harness/connector/mappers/terraformcloudmapper/TerraformCloudEntityToDTO.java#L36)):*
> ```Java
>             .executeOnDelegate(connector.getExecuteOnDelegate())
>             .delegateSelectors(connector.getDelegateSelectors());
>     final TerraformCloudCredentialType credentialType = connector.getCredentialType();
> ```

Looks like e don't need to create a new variable for credentialType.
___
*[`950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java` line 453 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDIeg91cgz6Kb_sLjW:-NMDIeg91cgz6Kb_sLjX:b52d9j7) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java#L453)):*
> ```Java
>   AWS_ASG_BLUE_GREEN_ROLLBACK_TASK_NG(TaskGroup.ASG, "AWS Asg Blue Green Rollback"),
>   SCM_BATCH_GET_FILE_TASK(TaskGroup.GIT, "SCM Batch Get File Task"),
>   TERRAFORM_CLOUD_TASK_NG(TaskGroup.TERRAFORM_CLOUD, "Terraform Cloud Task NG");
> ```

let's add unsuported: true, we will enable it only after everything will be ready.
___
*[`954-connector-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloudconnector/TerraformCloudCredentialType.java` line 26 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDK6Zo2kIRQuHUkBMr:-NMDK6Zo2kIRQuHUkBMs:bbyrhmp) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/954-connector-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloudconnector/TerraformCloudCredentialType.java#L26)):*
> ```Java
>   }
> 
>   public String getDisplayName() {
> ```

do we need public String getDisplayName() if we already have final String displayName() ?
___
*[`970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudConfig.java` line 22 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDLhx7Du89apusSUDH:-NMDLhx86opiB2whQk98:bhdgxhu) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudConfig.java#L22)):*
> ```Java
> @ToString(exclude = {"token"})
> public class TerraformCloudConfig {
>   String url;
> ```

Should we have here an interface and a specific implementation for URL , Token authentication (so another implementation can be easily added)?

**Review by alebuh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 56 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`440-connector-nextgen/src/main/java/io/harness/connector/mappers/terraformcloudmapper/TerraformCloudEntityToDTO.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDFksoBMpjdNpt6uEM:-NMDdAbIBg85DS1C0jt-:b499q7j) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/440-connector-nextgen/src/main/java/io/harness/connector/mappers/terraformcloudmapper/TerraformCloudEntityToDTO.java#L36)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Looks like e don't need to create a new variable for credentialType.
</blockquote></details>

Changed
___
*[`950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java` line 453 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDIeg91cgz6Kb_sLjW:-NMDdJ1s1pDlcVjy3DoB:b-7tfmjp) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java#L453)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

let's add unsuported: true, we will enable it only after everything will be ready.
</blockquote></details>

added
___
*[`954-connector-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloudconnector/TerraformCloudCredentialType.java` line 26 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDK6Zo2kIRQuHUkBMr:-NMDdLQz3TfwKDJzERez:b-hfpv0l) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/954-connector-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloudconnector/TerraformCloudCredentialType.java#L26)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

do we need public String getDisplayName() if we already have final String displayName() ?
</blockquote></details>

removed
___
*[`970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudConfig.java` line 22 at r2](https://reviewable.io/reviews/harness/harness-core/42813#-NMDLhx7Du89apusSUDH:-NMDdNTmAMKIjzyUmDdX:b-q2fq8d) ([raw file](https://github.com/harness/harness-core/blob/7d7345d8719a794ebc610eaa4cbf00b77f08944a/970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudConfig.java#L22)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Should we have here an interface and a specific implementation for URL , Token authentication (so another implementation can be easily added)?
</blockquote></details>

You are right, I added this just as a placeholder (didn't want to bring more files into PR ). It's not used anywhere. However, changed now. Added interface TerraformCloudCredentials which will be implemented by a different type of authentication. Once we need more configuration details in TerraformCloudConfig we will add fields like organisationId, workplaceId, configurationId.... and whatever we might need.

**Comment by alebuh** [comment]
trigger smartchecks

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloud/TerraformCloudValidationParams.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+
+@OwnedBy(HarnessTeam.CDP)
+@Data
+@SuperBuilder
+public class TerraformCloudValidationParams
+    extends ConnectorTaskParams implements ConnectorValidationParams, ExecutionCapabilityDemander {
+  TerraformCloudConnectorDTO terraformCloudConnectorDTO;
```
Comment (vlica-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloud/TerraformCloudValidationParams.java:30]: minor: I think access modifier should be private when we use @Data, AFAIK @Data from lombok does not make it private

**Review by vlica-harness — DISMISSED** [review]

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloud/TerraformCloudValidationParams.java (hunk 1 above)
Comment (alebuh) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/terraformcloud/TerraformCloudValidationParams.java:30]: You are right, @DaTa makes it public. The most of ValidationParams classes have this annotation so I followed that pattern. My assumption is that is just added for code reduction so we don't have to implement getters and setters

**Comment by alebuh** [comment]
trigger feature-build

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 14 of 15 files at r3, 1 of 2 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 58 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 37 of 47 files at r1, 3 of 4 files at r2, 14 of 15 files at r3, 1 of 2 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 1. unresolved discussions.

**Comment by alebuh** [comment]
trigger ft1

**Comment by alebuh** [comment]
trigger ft2

**Comment by alebuh** [comment]
trigger ft1

**Comment by alebuh** [comment]
trigger ti1

**Review by tmacari-harness — DISMISSED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by vlica-harness — DISMISSED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 1. unresolved discussions.

**Review by mdmilic — APPROVED** [review]
Approving delegate owned file changes

> Reviewed 6 of 47 files at r1, 5 of 15 files at r3, 1 of 2 files at r4, 1 of 2 files at r5, 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813#-:-NMUimyB4Bq2HDr_5fQ2:b-sjeihy)* status: 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by alebuh** [comment]
trigger ti0

**Comment by raghuAtWings** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskNG.java` line 51 at r7](https://reviewable.io/reviews/harness/harness-core/42813#-NMV2lWZ0h3pxgkrznyJ:-NMV2lWZ0h3pxgkrznyK:bve050m) ([raw file](https://github.com/harness/harness-core/blob/d62b5ccf0a94971d4ae62a451fe4b79c525b7720/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskNG.java#L51)):*
> ```Java
>   @Override
>   public DelegateResponseData run(TaskParameters parameters) {
>     log.info("Started executing Terraform Cloud Task NG");
> ```

do you need this log line?

**Comment by raghuAtWings** [comment]
*[`970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudApiTokenCredentials.java` line 21 at r7](https://reviewable.io/reviews/harness/harness-core/42813#-NMV3GYS91xaq6TXLf59:-NMV3GYS91xaq6TXLf5A:b-7fv9i2) ([raw file](https://github.com/harness/harness-core/blob/d62b5ccf0a94971d4ae62a451fe4b79c525b7720/970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudApiTokenCredentials.java#L21)):*
> ```Java
> @Builder
> @ToString(exclude = {"token"})
> public class TerraformCloudApiTokenCredentials implements TerraformCloudCredentials {
> ```

why does this need to be in 970? can it move to upper layers?

**Review by raghuAtWings — COMMENTED** [review]
> Reviewed 37 of 47 files at r1, 3 of 4 files at r2, 13 of 15 files at r3, 1 of 2 files at r4, 2 of 2 files at r5, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by alebuh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 3 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskNG.java` line 51 at r7](https://reviewable.io/reviews/harness/harness-core/42813#-NMV2lWZ0h3pxgkrznyJ:-NMY3kUD52SNb7AXHUyT:bgahv3) ([raw file](https://github.com/harness/harness-core/blob/d62b5ccf0a94971d4ae62a451fe4b79c525b7720/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskNG.java#L51)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

do you need this log line?
</blockquote></details>

it's not necessary. Removed! 
___
*[`970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudApiTokenCredentials.java` line 21 at r7](https://reviewable.io/reviews/harness/harness-core/42813#-NMV3GYS91xaq6TXLf59:-NMY4Rj76e5BeQW6HqV3:bgbdwmb) ([raw file](https://github.com/harness/harness-core/blob/d62b5ccf0a94971d4ae62a451fe4b79c525b7720/970-api-services-beans/src/main/java/io/harness/terraformcloud/TerraformCloudApiTokenCredentials.java#L21)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why does this need to be in 970? can it move to upper layers?
</blockquote></details>

This is used in the 930-delegate-task and 932-connector-task modules. I thought this was the right place to be added as it how was done with AwsConfig, AzureConfig, SshSessionConfig, CloudFoundryConfig... Do you have any suggestion on where should we move this or it's good to stay here?

**Comment by alebuh** [comment]
trigger feature-build

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 7 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 3. unresolved discussions.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42813)* status: 1. unresolved discussions.

**Pull request merged**